text/0000-time.md

```diff
@@ -0,0 +1,359 @@
+}
+```
+The `value` is a `Duration` type, and thus, `ClockInfo` can be thought of as a POSIX `timespec`, but much better. It provides descriptions of what is being read, not just the value. This additional metadata is essential for scientific applications that want to use time as a tool for measuring some event. Applications need to know about the properties of clocks to get accurate measurements and reduce errors in experiments.
+
+`ClockUsage` is a set of enums describing the purpose of the clock. This gives users control of what type of clock they are reading from (when talking to the kernel), and they can expect certain invariants to be maintained, such as a clock being monotonic.
+
+```rust
+enum ClockUsage {
+    Monotonic,
+    RealTime,
+    Unknown
+}
+```
+
+The kernel internally manages a list of usable clocks backed by hardware. The `id` identifies which clock source is used when interacting with the `Clock`. User programs can get time in a variety of ways. Either transparently using Rust's `std::time` crate, directly through system calls, or indirectly through methods exposed by the `Clock` type. 
+
+## System Call Interface
+
+Revisiting our example from earlier, let's see how it would work when performing a system call:
+```rust
+use crate twizzler_abi::syscall::sys_read_clock_info;
+use crate clock::{ClockInfo, ReadClockFlags};
+
+fn main() {
+    // reads a moment in time
+    // returns a ClockInfo type
+    let now = sys_read_clock_info(ClockUsage::Monotonic, 0, ReadClockFlags::empty());
+    
+    // elapsed time since now
+    let later = sys_read_clock_info(ClockUsage::Monotonic, 0, ReadClockFlags::empty());
+    // ClockInfo.value() returns a Duration type
+    let elapsed_time = later.value() - now.value();
+    println!("Running main() took {} nanoseconds.", elapsed_time.as_nanos());
+}
+```
+A few things here. For starters, `sys_read_clock_info`, a new system call. We pass in the type of clock we want (`ClockUsage`), an optional `clock_id`, and some flags. By default, this returns a `ClockInfo` object with all fields filled in. The returned `ClockInfo`, in this case, is generated from a `ClockUsage` specific clock. We will get to the use of `clock_id` later.
+
+Semantically, this is the same program, except it uses different abstractions. The `ClockInfo` type has a set of methods to return internal values. We calculate the elapsed time by subtracting two `Duration` types that sampled different points in time from the same clock. Internally, `std::time` does something like this using the `Instant` type.
+
+Twizzler exposes a new set of system calls related to timekeeping. Other than `sys_read_clock_info`, which is helpful in reading a clock and learning about its properties, users need ways to steer a clock and prevent drift. Users also need support to discover available clocks.
+
+```rust
+fn sys_read_clock_info(clock: ClockUsage, clock_id: u64, flags: ReadClockFlags) -> Result<ClockInfo, ReadClockError>
```

Comment:
As we discussed on wednesday, I'd like to see this changed to an API that doesn't admit erroneous combinations of arguments.